sdk/core/azure-core/inc/uuid.hpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <new> // for placement new
+#include <string>
+#include <utility> // for swap and move
+#include <random>
+
+namespace Azure { namespace Core {
+
+  class UUID {
+
+  private:
+    static const int UUIDSize = 16;
+
+    uint8_t m_uuid[UUIDSize];
+    // The UUID reserved variants.
+    static constexpr uint8_t ReservedNCS = 0x80;
+    static constexpr uint8_t ReservedRFC4122 = 0x40;
+    static constexpr uint8_t ReservedMicrosoft = 0x20;
+    static constexpr uint8_t ReservedFuture = 0x00;
+
+  public:
+    UUID()
+    {
+      std::random_device rd;
+      std::mt19937 gen(rd());
+
+      for (int i = 0; i < UUIDSize; i+=4)
+        *reinterpret_cast<uint32_t*>(m_uuid + i) = gen();
```

Comment:
(suggestion) I would
```
for (auto i = 0; i < sizeof(m_uuid); ++i)
{
    m_uuid[i] = static_cast<decltype(m_uuid[0])>(gen() % std::numeric_limits<decltype(m_uuid[0])>::max);
}
```

In addition to that, I would change m_uuid to `std::array`, and use `m_uuid.size()` instead of `sizeof(m_uuid)`, and then use `decltype(m_uuid)::value_type` instead of `decltype(m_uuid[0])`.
`std::array` is exactly the plain C array after it is compiled, but it never decays to a pointer. `size()` is constexpr, so it takes no space and is substituted by int literal during compilation.

Reply:
I doubt this is the right way to use random engine in C++11. Does the C++ standard state that `gen()` returns a random number from 0~UINT32MAX?

Reply:
Reference to this example https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution

Reply:
http://www.cplusplus.com/reference/random/mt19937/ is `"A Mersenne Twister pseudo-random generator of 32-bit numbers"`, `mt19937::result_type` is `uint_fast32_t`, which is "The type of the numbers generated."

Reply:
okay, I was wrong. This also works.

Reply:
But still, you suggested code won't work. `std::numeric_limits<decltype(m_uuid[0])>::max` is 255. You can only get value between 0 and 254, both inclusive. You'll never get 255.

Reply:
But I get where you're coming from and I can understand if a cryptographer can argue that `%`-ing a generated random 32-bit number may not produce an even distribution; it's just that I can't understand that difference using my common logic. Maybe there is a difference. I liked the example you posted, with using `uniform_int_distribution` which does what I was trying to achieve with `%`, but ensures it is done in a civilized way.
```
#include <random>
#include <iostream>
 
int main()
{
    std::random_device rd;  //Will be used to obtain a seed for the random number engine
    std::mt19937 gen(rd()); //Standard mersenne_twister_engine seeded with rd()
    std::uniform_int_distribution<> distrib(1, 6);
 
    for (int n=0; n<10; ++n)
        //Use `distrib` to transform the random unsigned int generated by gen into an int in [1, 6]
        std::cout << distrib(gen) << ' ';
    std::cout << '\n';
}
```
So, in our case, we would need numbers from 0 to 255 (or `std::numeric_limits<decltype(m_uuid[0])>::min` to `std::numeric_limits<decltype(m_uuid[0])>::max`, to be exact).